### PR TITLE
Replace release.py with OIDC publishing

### PR DESCRIPTION
**Pull request by alex**
We are now in the PyPI OIDC publishing beta

### CODE REVIEW

**Comment by alex** [comment]
When this is merged it needs to be configured at https://pypi.org/manage/project/cryptography/settings/publishing/

**Comment by alex** [comment]
Plan of action:

- [ ] Add support for `workflow_dispatch` and publishing to Test PyPI
- [ ] Merge
- [ ] Do a test run publishing to Test PyPI, rinse and repeat until successful
- [ ] Configure on PyPI.

File: .github/workflows/pypi-publish.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+          response = requests.post(f"https://{os.environ['PYPI_DOMAIN']}/_/oidc/github/mint-token", json={"token": token})
+          response.raise_for_status()
+          pypi_token = response.json()["token"]
+
+          with open(os.environ["GITHUB_ENV"], "a") as f:
+              f.write("TWINE_PASSWORD={pypi_token}\n")
+        shell: python
+
+      - run: "twine upload --repository $TWINE_REPO dist/"
```
Comment (woodruffw) [inline .github/workflows/pypi-publish.yml:66]: Does `twine upload dist/` work? I've always thought `twine` needs to take each dist file individually, e.g. `dist/*` to glob them out.

Comment (alex) [reply .github/workflows/pypi-publish.yml:66]: Hahaha, probably it doesn't work. This is why testing is required!

Comment (woodruffw) [reply .github/workflows/pypi-publish.yml:66]: Docs imply that the input needs to be a file, but maybe this is undocumented: https://twine.readthedocs.io/en/stable/

**Comment by di** [comment]
@alex, What's your TestPyPI username? I'll need to put you in the beta there too but don't see an `alexgaynor` user.

**Comment by alex** [comment]
@di I didn't have one, I've just created one with the same name (alexgaynor).

(I also don't currently own the "cryptography" package on Test PyPI, but I've emailed the current owner to see if I can get it.)

**Comment by alex** [comment]
Swell, thanks!

**Comment by alex** [comment]
Test PyPI is now configured for this workflow, just need a review from @reaperhulk

File: .github/workflows/pypi-publish.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+          echo "PYPI_DOMAIN=test.pypi.org" >> GITHUB_ENV
+          echo "TWINE_REPO=testpypi" >> GITHUB_ENV
+        if: github.event_name == 'workflow_dispatch' && github.event.inputs.environment == 'testpypi'
+
+      - run: |
+          import os
+
+          import requests
+
+          response = requests.get(
```
Comment (woodruffw) [inline .github/workflows/pypi-publish.yml:51]: Tiny nitpick, but if you want a slightly cleaner param here you can pass the query args separately:

```python
params={"audience": os.environ["OIDC_AUDIENCE"]}
```

**Review by reaperhulk — APPROVED** [review]
let the experiment begin

**Comment by webknjaz** [comment]
@alex any reason you didn't go for https://github.com/marketplace/actions/pypi-publish#trusted-publishing that embeds the OIDC logic?

**Comment by webknjaz** [comment]
Ah.. you made this 5 days before https://github.com/pypa/gh-action-pypi-publish/pull/123 got merged.

**Pull request merged**